core/src/main/java/hudson/util/io/TarArchiver.java

```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
         int mode = IOUtils.mode(file);
         if (mode!=-1)   te.setMode(mode);
         te.setModTime(file.lastModified());
-        if(!file.isDirectory())
-            te.setSize(file.length());
-
-        tar.putArchiveEntry(te);
+        long size = 0;
 
         if (!file.isDirectory()) {
-            FileInputStream in = new FileInputStream(file);
-            try {
-                int len;
-                while((len=in.read(buf))>=0)
-                    tar.write(buf,0,len);
-            } finally {
-                in.close();
+            size = file.length();
+            te.setSize(size);
+        }
+        tar.putArchiveEntry(te);
+        try {
+            if (!file.isDirectory()) {
+                FileInputStream in = new FileInputStream(file);
+                try {
+                    int len;
+                    int read = 0;
+                    while ((len = in.read(buf)) >= 0) {
+                        if (len + read <= size) { // ensure we don't write more bytes than the declared when we created
+                                                  // the entry
+                            tar.write(buf, 0, len);
+                            read += len;
+                        } else {
+                            tar.write(buf, 0, (int) size - read);
+                            break;
+                        }
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
```

Comment:
Actually the exception is thrown on the final of the caller not on this one
(the reference I put before) and only when we don't have the final to close
the entry. I'll convert it to try with resources, and see if we can use
Jesse's suggestion.
El 22/6/2016 20:13, "Jesse Glick" <notifications@github.com> escribió:

> In core/src/main/java/hudson/util/io/TarArchiver.java
> <https://github.com/jenkinsci/jenkins/pull/2419#discussion_r68105507>:
>
> > +                FileInputStream in = new FileInputStream(file);
> > +                try {
> > +                    int len;
> > +                    int read = 0;
> > +                    while ((len = in.read(buf)) >= 0) {
> > +                        if (len + read <= size) { // ensure we don't write more bytes than the declared when we created
> > +                                                  // the entry
> > +                            tar.write(buf, 0, len);
> > +                            read += len;
> > +                        } else {
> > +                            tar.write(buf, 0, (int) size - read);
> > +                            break;
> > +                        }
> > +                    }
> > +                } catch (IOException e) {// log the exception in any case
> > +                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
>
> Sure, if the code idiom *can* be converted to try-with-resources, then by
> all means do it.
>
> BTW some Java IDEs offer this conversion as a hint.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/2419/files/64b039bc7a79590877c1800a9aa1bb23549a214f#r68105507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AQXNBPHtA-FuIO2fKadsRtDHxnCRfZzIks5qOXtAgaJpZM4I72ij>
> .
>